cpp/src/arrow/compute/exec/options.h

```diff
@@ -361,6 +361,19 @@ class ARROW_EXPORT HashJoinNodeOptions : public ExecNodeOptions {
   Expression filter;
 };
 
+class ARROW_EXPORT AsofJoinNodeOptions : public ExecNodeOptions {
+ public:
+  AsofJoinNodeOptions(FieldRef time, FieldRef keys, int64_t tolerance)
+      : time(std::move(time)), keys(std::move(keys)), tolerance(tolerance) {}
+
+  // time column
```

Comment:
There are quite a few restrictions on these options right now.  `time` must be int64?  `key` must be an int64?  The payload columns must meet a pretty restrictive set of types.

I think this is ok as this is an early implementation.  However, we should thoroughly document these limitations as well as test for these when the node is created.

Reply:
I agree these restrictions are not ideal. I put "Handle more datatypes (both key and value)" as follow ups only in the interest of time.

> However, we should thoroughly document these limitations as well as test for these when the node is created.
I agree. I can add tests for behaviors for when unsupported types are passed in - do you have a pointer to some existing tests that test similar behaviors?

Reply:
I added tests for the unsupported types.